### ISSUE TITLE
Delete has_one join_table row fails

### ISSUE THREAD

**Issue by benoittgt** [rails/rails#31198]
### Steps to reproduce

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.1.0"
  gem 'pg'
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: 'testing',
  schema_search_path: 'public'
)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true

  create_table :comments, force: true

  create_join_table :posts, :comments, force: true
end

class Post < ActiveRecord::Base
  has_one :comments_post, dependent: :destroy
  has_one :comment, through: :comments_post
end

class CommentsPost < ActiveRecord::Base
  belongs_to :post
  belongs_to :comment
end

class Comment < ActiveRecord::Base
  has_many :comments_posts
  has_many :post, through: :comments_posts
end

class BugTest < Minitest::Test
  def test_has_one_association
    post = Post.create!(comment: Comment.create!)
    assert_equal 1, Comment.count
    assert_equal 1, CommentsPost.count
    assert_equal 1, Post.count

    post.destroy!
    assert_equal 0, Post.count
    assert_equal 0, CommentsPost.count
    assert_equal 1, Post.count
  end
end
```

### Expected behavior

`post.destroy!` should delete `post` and the row in join table with the `id` of `post`.

### Actual behavior

It crashes.
On `4.2.1`
```
-- create_table(:posts, {:force=>true})
The PGconn, PGresult, and PGError constants are deprecated, and will be
removed as of version 1.0.

You should use PG::Connection, PG::Result, and PG::Error instead, respectively.

Called from /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-4.2.10/lib/active_support/dependencies.rb:240:in `load_dependency'
D, [2017-11-22T12:12:00.129445 #20409] DEBUG -- :    (7.4ms)  DROP TABLE "posts"
D, [2017-11-22T12:12:00.137379 #20409] DEBUG -- :    (7.6ms)  CREATE TABLE "posts" ("id" serial primary key)
   -> 0.0473s
-- create_table(:comments, {:force=>true})
D, [2017-11-22T12:12:00.141731 #20409] DEBUG -- :    (3.0ms)  DROP TABLE "comments"
D, [2017-11-22T12:12:00.149636 #20409] DEBUG -- :    (7.6ms)  CREATE TABLE "comments" ("id" serial primary key)
   -> 0.0122s
-- create_join_table(:posts, :comments, {:force=>true})
D, [2017-11-22T12:12:00.155581 #20409] DEBUG -- :    (2.1ms)  DROP TABLE "comments_posts"
D, [2017-11-22T12:12:00.157885 #20409] DEBUG -- :    (2.0ms)  CREATE TABLE "comments_posts" ("post_id" integer NOT NULL, "comment_id" integer NOT NULL)
   -> 0.0082s
Run options: --seed 60774

# Running:

D, [2017-11-22T12:12:00.197462 #20409] DEBUG -- :    (0.2ms)  BEGIN
D, [2017-11-22T12:12:00.200777 #20409] DEBUG -- :   SQL (0.5ms)  INSERT INTO "comments" DEFAULT VALUES RETURNING "id"
D, [2017-11-22T12:12:00.202300 #20409] DEBUG -- :    (1.1ms)  COMMIT
D, [2017-11-22T12:12:00.219642 #20409] DEBUG -- :    (0.1ms)  BEGIN
D, [2017-11-22T12:12:00.220549 #20409] DEBUG -- :   SQL (0.4ms)  INSERT INTO "posts" DEFAULT VALUES RETURNING "id"
D, [2017-11-22T12:12:00.222413 #20409] DEBUG -- :   SQL (0.2ms)  INSERT INTO "comments_posts" ("comment_id", "post_id") VALUES ($1, $2)  [["comment_id", 1], ["post_id", 1]]
D, [2017-11-22T12:12:00.224116 #20409] DEBUG -- :    (1.1ms)  COMMIT
D, [2017-11-22T12:12:00.226754 #20409] DEBUG -- :    (0.5ms)  SELECT COUNT(*) FROM "comments"
D, [2017-11-22T12:12:00.228491 #20409] DEBUG -- :    (0.5ms)  SELECT COUNT(*) FROM "comments_posts"
D, [2017-11-22T12:12:00.230018 #20409] DEBUG -- :    (0.3ms)  SELECT COUNT(*) FROM "posts"
D, [2017-11-22T12:12:00.230346 #20409] DEBUG -- :    (0.1ms)  BEGIN
D, [2017-11-22T12:12:00.231393 #20409] DEBUG -- :    (0.1ms)  ROLLBACK
E

Finished in 0.041619s, 24.0275 runs/s, 72.0825 assertions/s.

  1) Error:
BugTest#test_has_one_association:
ActiveRecord::StatementInvalid: PG::SyntaxError: ERROR:  zero-length delimited identifier at or near """"
LINE 1: DELETE FROM "comments_posts" WHERE "comments_posts"."" = $1
                                                            ^
: DELETE FROM "comments_posts" WHERE "comments_posts"."" = $1
```

On `5.1.0`

```
  1) Error:
BugTest#test_has_one_association:
NoMethodError: undefined method `to_sym' for nil:NilClass
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/sanitization.rb:90:in `block in expand_hash_conditions_for_aggregates'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/sanitization.rb:89:in `each'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/sanitization.rb:89:in `expand_hash_conditions_for_aggregates'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/relation/where_clause_factory.rb:15:in `build'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/relation/query_methods.rb:612:in `where!'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/relation/query_methods.rb:605:in `where'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/persistence.rb:558:in `relation_for_destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/locking/optimistic.rb:127:in `relation_for_destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/persistence.rb:554:in `destroy_row'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/counter_cache.rb:191:in `destroy_row'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/locking/optimistic.rb:117:in `destroy_row'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/persistence.rb:195:in `destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/callbacks.rb:321:in `block in destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:97:in `run_callbacks'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:825:in `_run_destroy_callbacks'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/callbacks.rb:321:in `destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:303:in `block in destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:384:in `block in with_transaction_returning_status'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/connection_adapters/abstract/database_statements.rb:223:in `transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:210:in `transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:381:in `with_transaction_returning_status'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:303:in `destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/associations/has_one_association.rb:59:in `delete'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/associations/has_one_association.rb:20:in `handle_dependency'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/associations/builder/association.rb:140:in `block in add_destroy_callbacks'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:413:in `instance_exec'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:413:in `block in make_lambda'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:197:in `block (2 levels) in halting'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:601:in `block (2 levels) in default_terminator'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:600:in `catch'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:600:in `block in default_terminator'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:198:in `block in halting'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:507:in `block in invoke_before'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:507:in `each'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:507:in `invoke_before'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:130:in `run_callbacks'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activesupport-5.1.0/lib/active_support/callbacks.rb:825:in `_run_destroy_callbacks'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/callbacks.rb:321:in `destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:303:in `block in destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:384:in `block in with_transaction_returning_status'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/connection_adapters/abstract/database_statements.rb:225:in `block in transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/connection_adapters/abstract/transaction.rb:194:in `block in within_new_transaction'
    /Users/bti/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/connection_adapters/abstract/transaction.rb:191:in `within_new_transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/connection_adapters/abstract/database_statements.rb:225:in `transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:210:in `transaction'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:381:in `with_transaction_returning_status'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/transactions.rb:303:in `destroy'
    /Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-5.1.0/lib/active_record/persistence.rb:211:in `destroy!'
    has_one_through.rb:65:in `test_has_one_association'

1 runs, 3 assertions, 0 failures, 1 errors, 0 skips
```

### System configuration
**Activerecord version**:   "5.1.0" and "4.2.1"
**Ruby version**: 2.4.1

Do I failed somewhere?

**Comment by benoittgt** [comment]
I think this is sadly related to https://github.com/rails/rails/issues/25347

**Comment by benoittgt** [comment]
It works with `has_and_belongs_to_many`

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "4.2.1"
  gem 'pg'
  gem 'pry'
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: 'testing',
  schema_search_path: 'public'
)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true

  create_table :comments, force: true

  create_join_table :posts, :comments, force: true
end

class Post < ActiveRecord::Base
  has_and_belongs_to_many :comments
end

class Comment < ActiveRecord::Base
  has_and_belongs_to_many :posts
end

class BugTest < Minitest::Test
  def test_has_one_association
    post = Post.create!(comments: [Comment.create!])
    assert_equal 1, Comment.count
    # assert_equal 1, CommentsPost.count
    assert_equal 1, Post.count

    post.destroy!
    assert_equal 1, Comment.count
    assert_equal 0, Post.count
    # assert_equal 0, CommentsPost.count
  end
end
```

```
....
# Running:

D, [2017-11-22T15:19:18.001644 #23598] DEBUG -- :    (0.1ms)  BEGIN
D, [2017-11-22T15:19:18.005025 #23598] DEBUG -- :   SQL (0.5ms)  INSERT INTO "comments" DEFAULT VALUES RETURNING "id"
D, [2017-11-22T15:19:18.007072 #23598] DEBUG -- :    (1.3ms)  COMMIT
/Users/bti/.rvm/gems/ruby-2.4.1/gems/activerecord-4.2.1/lib/active_record/associations/collection_association.rb:267: warning: constant ::Fixnum is deprecated
D, [2017-11-22T15:19:18.038416 #23598] DEBUG -- :    (0.3ms)  BEGIN
D, [2017-11-22T15:19:18.041797 #23598] DEBUG -- :   SQL (1.5ms)  INSERT INTO "posts" DEFAULT VALUES RETURNING "id"
D, [2017-11-22T15:19:18.045204 #23598] DEBUG -- :   SQL (0.3ms)  INSERT INTO "comments_posts" ("comment_id", "post_id") VALUES ($1, $2)  [["comment_id", 1], ["post_id", 1]]
D, [2017-11-22T15:19:18.048122 #23598] DEBUG -- :    (1.2ms)  COMMIT
D, [2017-11-22T15:19:18.048884 #23598] DEBUG -- :    (0.2ms)  SELECT COUNT(*) FROM "comments"
D, [2017-11-22T15:19:18.049359 #23598] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "posts"
D, [2017-11-22T15:19:18.049554 #23598] DEBUG -- :    (0.1ms)  BEGIN
D, [2017-11-22T15:19:18.050296 #23598] DEBUG -- :   SQL (0.1ms)  DELETE FROM "comments_posts" WHERE "comments_posts"."post_id" = $1  [["post_id", 1]]
D, [2017-11-22T15:19:18.050839 #23598] DEBUG -- :   SQL (0.1ms)  DELETE FROM "posts" WHERE "posts"."id" = $1  [["id", 1]]
D, [2017-11-22T15:19:18.052189 #23598] DEBUG -- :    (1.2ms)  COMMIT
D, [2017-11-22T15:19:18.052937 #23598] DEBUG -- :    (0.2ms)  SELECT COUNT(*) FROM "comments"
D, [2017-11-22T15:19:18.053490 #23598] DEBUG -- :    (0.2ms)  SELECT COUNT(*) FROM "posts"
.

Finished in 0.056713s, 17.6326 runs/s, 70.5306 assertions/s.

1 runs, 4 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by benoittgt** [comment]
TL;DR don't use `create_join_table`.

Working version : 

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "4.2.1"
  gem 'pg'
  gem 'pry'
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: 'testing',
  schema_search_path: 'public'
)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true

  create_table :comments, force: true

  create_table :comments_posts, force: true do |t|
    t.integer  "comment_id"
    t.integer  "post_id"
  end
end

class Post < ActiveRecord::Base
  has_one :comments_post, dependent: :destroy
  has_one :comment, through: :comments_post
end

class CommentsPost < ActiveRecord::Base
  belongs_to :post
  belongs_to :comment
end

class Comment < ActiveRecord::Base
  has_many :comments_posts
  has_and_belongs_to_many :posts, through: :comments_posts
end

class BugTest < Minitest::Test
  def test_has_one_association
    post = Post.create!(comment: Comment.create!)

    assert_equal 1, Comment.count
    assert_equal 1, CommentsPost.count
    assert_equal 1, Post.count

    post.destroy!

    assert_equal 1, Comment.count
    assert_equal 0, Post.count
    assert_equal 0, CommentsPost.count
  end
end
```

It works on `4.2.1` and `5.1.0`.

```
SQL (0.4ms)  DELETE FROM "comments_posts" WHERE "comments_posts"."id" = $1  [["id", 1]]
SQL (0.3ms)  DELETE FROM "posts" WHERE "posts"."id" = $1  [["id", 1]]
```
